fms_extras/utils/generation.py

```diff
@@ -3,12 +3,344 @@
 from typing import Any, Callable, Dict, List, MutableMapping, Optional, Tuple, Union
 
 import torch
+import torch.nn as nn
 import torch.nn.functional as F
 
 from fms_extras.models.speculator import MLPSpeculator, apply_index_map, flatten_batch
 from fms_extras.utils.cache.paged import PagedAttentionCacheData, PagedKVCacheManager
 
 
+def __prepare_inputs_for_prefill(
+    input_ids: Union[torch.Tensor, List[torch.Tensor]],
+    kv_cache_manager: PagedKVCacheManager,
+    model_input_lengths: List[int],
+    max_length_in_batch: int,
+    decode_seq_length: int,
+    max_sequence_length_in_model: int,
+) -> Tuple[torch.Tensor, PagedAttentionCacheData, torch.Tensor]:
+    """Prepare all inputs required to perform prefill step on prompt tokens
+
+    This function will:
+
+    Args:
+        input_ids: Union[torch.Tensor, List[torch.Tensor]]
+            list of prompts where each tensor in the list corresponds to a single prompt in the batch
+        kv_cache_manager: PagedKVCacheManager
+            the paged attention kv-cache manager
+        model_input_lengths: List[int]
```

Comment:
Oh I see, we're calculating and re-using these in `speculative_generate`. I'm inclined to recalculate `model_input_lengths` and `max_length_in_batch` here, and have the two lines of redundant code to avoid passing two extraneous args, but nbd either way

Reply:
yes, I was trying to reuse it. but its small enough that it probably doesn't matter